libs/play/index.js

```diff
@@ -339,29 +339,49 @@ function playSubdomain(hostname) {
 export async function handleRunner(req, res) {
   const url = new URL(req.url, "https://example.com");
   if (url.searchParams.has("blank")) {
-    return res.setHeader("Content-Type", "text/html").status(200).end();
+    res.setHeader("Content-Type", "text/html").status(204).end();
+    return;
   }
   const referer = new URL(
     req.headers["referer"] || "https://example.com",
     "https://example.com"
   );
   const stateParam = url.searchParams.get("state");
+
+  if (!stateParam) {
+    console.warn("[runner] Missing state parameter");
+    res.status(400).end();
+    return;
+  }
+
   const { state, hash } = await decompressFromBase64(stateParam);
 
-  const isLocalhost = req.hostname === "localhost";
-  const hasMatchingHash = playSubdomain(req.hostname) === hash;
-  const isIframeOnMDN =
-    referer.hostname === ORIGIN_MAIN &&
-    req.headers["sec-fetch-dest"] === "iframe";
-
-  if (
-    !stateParam ||
-    !state ||
-    (!isLocalhost && !hasMatchingHash && !isIframeOnMDN)
-  ) {
+  if (!state) {
+    console.warn("[runner] Invalid state parameter");
     return res.status(404).end();
   }
 
+  if (req.hostname !== "localhost") {
+    const expectedHash = playSubdomain(req.hostname);
+
+    if (expectedHash !== hash) {
+      console.warn(
+        `[runner] Hash mismatch: ${JSON.stringify({ expectedHash, hash })}`
+      );
```

Comment:
This is gonna log for every single play request from prod so we should swap the order with checking the conditions below, and only log either when we actually send the response

Reply:
Removed in e12f556226590f59c7ff3c3f5ab26b8aa4f73540 as part of refinement, to make the errors more insightful.